test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ClientTests/Tests/ClientQueryTests.cs

```diff
@@ -0,0 +1,228 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData.Client.E2E.TestCommon;
+using Microsoft.OData.Client.E2E.Tests.ClientTests.Server;
+using Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd;
+using Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.Default;
+using Microsoft.OData.Client.E2E.Tests.Common.Server.EndToEnd;
+using Xunit;
+
+namespace Microsoft.OData.Client.E2E.Tests.ClientTests.Tests
+{
+    public class ClientQueryTests : EndToEndTestBase<ClientQueryTests.TestsStartup>
+    {
+        private readonly Uri _baseUri;
+        private readonly Container _context;
+
+        public class TestsStartup : TestStartupBase
+        {
+            public override void ConfigureServices(IServiceCollection services)
+            {
+                services.ConfigureControllers(typeof(ClientQueryTestsController), typeof(MetadataController));
+
+                services.AddControllers().AddOData(opt => opt.EnableQueryFeatures()
+                    .AddRouteComponents("odata", CommonEndToEndEdmModel.GetEdmModel()));
+            }
+        }
+
+        public ClientQueryTests(TestWebApplicationFactory<TestsStartup> fixture)
+            : base(fixture)
+        {
+            _baseUri = new Uri(Client.BaseAddress, "odata/");
+            _context = new Container(_baseUri);
+            _context.HttpClientFactory = HttpClientFactory;
+        }
+
+        [Fact]
+        public void DollarFilter_UsingContains_ExecutesSuccessfully()
+        {
+            var testlists = new List<string>()
+            {
+                new("People?$filter=contains(Name, 'm')"),
+                new("People?$filter=contains('m', Name)"),
+                new("People?$filter=not contains(Name, 'm')"),
+                new("People?$filter=contains(Name, 'm') eq true"),
+                new("People?$filter=false eq contains(Name, 'm')"),
+                new("People?$filter=contains(Name, substring('name',  2, 1))"),
+                new("People?$filter=contains(concat('User','name'), 'name')"),
+                new("People?$filter=contains(concat('User','name'), substring('name',  2, 1))"),
+            };
+
+            foreach (var test in testlists)
+            {
+                var result = _context.Execute<Common.Clients.EndToEnd.Person>(new Uri(_baseUri.OriginalString + test)).ToArray();
+                Assert.NotNull(result);
+            }
+        }
+
+        [Fact]
+        public void Using_LinqContains_ExecutesSuccessfully()
+        {
+            var result = (from c in _context.People
+                          where c.Name.Contains("m")
+                          select new Common.Clients.EndToEnd.Person() { Name = c.Name }) as DataServiceQuery<Common.Clients.EndToEnd.Person>;
+
+            Assert.NotNull(result);
+
+            result = (DataServiceQuery<Common.Clients.EndToEnd.Person>)_context.People.Where(c => c.Name.Contains("m"));
+            Assert.NotNull(result);

```

Comment:
I don't think `result` can ever be null in this case. If the cast fails, an exception is thrown. If you want to use null to test whether the cast failed, then use `as` instead.

But I think a better assertion would be to check the OData URL generated for this call.